zarr/tests/test_convenience.py

```diff
@@ -85,3 +90,516 @@ def test_lazy_loader():
+        dest = self.new_dest()
+
+        # copy group, dest array in the way
+        dest.create_dataset('foo/bar', shape=(10,))
+
+        # raise
+        with pytest.raises(CopyError):
+            copy(source['foo'], dest)
+        assert dest['foo/bar'].shape == (10,)
+        with pytest.raises(CopyError):
+            copy(source['foo'], dest, if_exists='raise')
+        assert dest['foo/bar'].shape == (10,)
+
+        # skip
+        copy(source['foo'], dest, if_exists='skip')
+        assert dest['foo/bar'].shape == (10,)
+
+        # replace
+        copy(source['foo'], dest, if_exists='replace')
+        check_copied_group(source['foo'], dest['foo'])
+
+    def test_copy_group_dry_run(self):
+        source = self.source
+        dest = self.new_dest()
+
+        # dry run, empty destination
+        n_copied, n_skipped, n_bytes_copied = \
+            copy(source['foo'], dest, dry_run=True, return_stats=True)
+        assert 0 == len(dest)
+        assert 3 == n_copied
+        assert 0 == n_skipped
+        assert 0 == n_bytes_copied
+
+        # dry run, array exists in destination
+        baz = np.arange(100, 200)
+        dest.create_dataset('foo/bar/baz', data=baz)
+        assert not np.all(source['foo/bar/baz'][:] == dest['foo/bar/baz'][:])
+        assert 1 == len(dest)
+
+        # raise
+        with pytest.raises(CopyError):
+            copy(source['foo'], dest, dry_run=True)
+        assert 1 == len(dest)
+
+        # skip
+        n_copied, n_skipped, n_bytes_copied = \
+            copy(source['foo'], dest, dry_run=True, if_exists='skip',
+                 return_stats=True)
+        assert 1 == len(dest)
+        assert 2 == n_copied
+        assert 1 == n_skipped
+        assert 0 == n_bytes_copied
+        assert_array_equal(baz, dest['foo/bar/baz'])
+
+        # replace
+        n_copied, n_skipped, n_bytes_copied = \
+            copy(source['foo'], dest, dry_run=True, if_exists='replace',
+                 return_stats=True)
+        assert 1 == len(dest)
+        assert 3 == n_copied
+        assert 0 == n_skipped
+        assert 0 == n_bytes_copied
+        assert_array_equal(baz, dest['foo/bar/baz'])
+
+    def test_logging(self):
+        source = self.source
+        dest = self.new_dest()
+
+        # callable log
+        copy(source['foo'], dest, dry_run=True, log=print)
+
+        # file name
+        fn = tempfile.mktemp()
+        atexit.register(os.remove, fn)
+        copy(source['foo'], dest, dry_run=True, log=fn)
+
+        # file
+        with tempfile.TemporaryFile(mode='w') as f:
+            copy(source['foo'], dest, dry_run=True, log=f)
+
+        # bad option
+        with pytest.raises(TypeError):
+            copy(source['foo'], dest, dry_run=True, log=True)
+
+
+try:
+    import h5py  # no qa
+    have_h5py = True
+except ImportError:  # pragma: no cover
+    have_h5py = False
+
+
+def temp_h5f():
+    fn = tempfile.mktemp()
+    atexit.register(os.remove, fn)
```

Comment:
Kind of a silly question, but should we be closing this first?

Reply:
Do you mean closing the HDF5 file at exit, before removing the file?

Reply:
Yep.

Reply:
Yes probably should. Easiest thing might be to call close() inside the test class tearDown rather than try to handle at exit.

Reply:
SGTM